### ISSUE TITLE
[Bug] Events are skipped on fast chains (Arbitrum)

### ISSUE THREAD

**Issue by 0xOlias** [0xOlias/ponder#288]
Because Ponder supports indexing multiple chains in the same app, it uses block timestamps to order events (rather than block numbers which are meaningless across chains). Ponder encounters a bug if a chain has several consecutive blocks with the same block timestamp (note that block timestamps are only granular to 1 second).

### Background

After ingesting a new head block for a given chain, Ponder's realtime sync service emits an event (`newRealtimeCheckpoint`) to downstream services announcing that logs are now available **up to and including** the block timestamp of the newly processed block.

Then, the user handler service runs a DB query to fetch these new logs from the store and runs them against the user-provided handler code. Internally, the user handler service stores a piece of mutable state `eventsProcessedToTimestamp`, again with a 1-second granularity.

### The problem

The bug occurs when a series of blocks is produced that looks like this.
```
block 10, timestamp 100
block 11, timestamp 100
------> ingest latest block (11) and emit newRealtimeCheckpoint(timestamp: 100)
  ----> user handlers service processes new logs, sets eventsProcessedToTimestamp: 100
block 12, timestamp 100 <-- logs in this block are never handled
block 13, timestamp 100 <-- logs in this block are never handled
block 14, timestamp 101
------> ingest latest block (14) and emit newRealtimeCheckpoint(timestamp: 101)
  ----> user handlers service processes new logs, sets eventsProcessedToTimestamp: 101
block 15, timestamp 101
...
```

### Potential solutions

One solution would be to update the realtime sync service to only emit `newRealtimeCheckpoint(timestamp)` once it has confirmed that no more blocks will be produced with the same timestamp.

A naive implementation would be `newRealtimeCheckpoint(timestamp: newHeadBlock.timestamp - 1)`. This effectively means: only process blocks once we've received the next block. On a fast chain like arbitrum this works pretty well - we're getting new blocks faster than 1/second. We're often ingesting & processing blocks in batches of ~4 anyway.

However, this doesn't work well on a chain like mainnet, because the service would always be one entire block (~12s) behind latest.

**Comment by AFDudley** [comment]
Two quick solutioms come to mind:
1. Can you just append the block number to the end of the time stamp?
2. If you know the endpoint is synced, them you can use the time you received the block instead of the timestamp.

The higher level problem is that generally speaking there no way to map "blockchain time" to "wall clock time", especially across multiple chains. Happy to discuss ways which we've addressed this problem over the years if 1. Isn't appealing.

**Comment by 0xOlias** [comment]
> The higher level problem is that generally speaking there no way to map "blockchain time" to "wall clock time", especially across multiple chains.

I'm generally aware of this problem, but haven't studied it super closely. Do you think it's ever reasonable to attempt to order events across multiple chains?

**Comment by 0xOlias** [comment]
Thanks a bunch for your input @AFDudley, that makes sense. Agree on the 2nd point. Supporting multiple chains in the same app is kind of an "experimental" feature; in practice nobody is doing it (yet). So, I think Ponder's current event ordering scheme:

1. Block timestamp
2. Chain ID
3. EVM execution order (Block number + Transaction index + Log index)

is reasonable and we can certainly revisit it.

### Proposed solution

With all that said, I think I have a workable solution to this bug described above.
1. In the `RealtimeService` setup method, estimate the connected network's average block time in seconds. This could be as simple as `(latest_block_timestamp - latest_minus_N_block_timestamp) / N` where N is `min(latest_block_number, 50)` to account for brand new chains (eg Anvil).
2. If `average_block_time < ~1.5 seconds`, switch the service to operate in "fast chain mode". In fast chain mode, the service only reports a given block timestamp (unix epoch) as ready/cached once it has ingested a block with a greater block timestamp. This introduces a small amount of delay.
3. If `average_block_time >= ~1.5 seconds`, operate the service in "normal mode". In normal mode, the service reports a given block timestamp as ready/cached as soon as the block has been ingested. There is no lag.

The reason we can't always operate in "fast chain mode" is because on a slow chain like mainnet, we would always be reporting new blocks one full block late (12 seconds). But it's probably acceptable to report blocks ~1 second late on fast chains like Arbitrum.

All said, not in love with this solution because it introduces lag. 

Thinking this through made me realize another issue for multiple chain apps. As Ponder is designed today, events are only handled up to the oldest/stalest chain. Consider an app that uses both mainnet and Optimism:

| Block timestamp | Chain    | Block number |
| --------------- | -------- | ------------ |
| 1000            | Mainnet  | 15           |
| 1001            | Optimism | 701          |
| 1002            | Optimism | 702          |
| ...             | Optimism | ...          |
| 1011            | Optimism | 711          |
| 1012            | Mainnet  | 16           |

While Optimism blocks 701-711 come in, the latest block timestamp on Mainnet is still 1000. Because Ponder currently only handles events up to the stalest chain, all of Optimism events 701-711 won't get handled until Mainnet block 16 in ingested. Womp! Will continue to think about this.

**Comment by SanLeo461** [comment]
Hi, is there any update on this bug?
I'm going to be doing some indexing on Arbitrum Nova and just saw this issue, hoping a fix isn't too far off! 🙂

**Comment by arberx** [comment]
> events are only handled up to the oldest/stalest chain. 

Trying to understand this a bit more. Whats the reason `Ponder` fires a `newCheckpoint` event at some point for all chains tracked, instead of per chain? 

e.g I imagine the `newCheckpoint` event handler to take in the chainId:
```
   this.syncGatewayService.on("newCheckpoint", async (chainId) => {
      await this.indexingService.processEvents(chainId);
    });

```

**Comment by SanLeo461** [comment]
> > events are only handled up to the oldest/stalest chain.
> 
> Trying to understand this a bit more. Whats the reason `Ponder` fires a `newCheckpoint` event at some point for all chains tracked, instead of per chain?
> 
> e.g I imagine the `newCheckpoint` event handler to take in the chainId:
> 
> ```
>    this.syncGatewayService.on("newCheckpoint", async (chainId) => {
>       await this.indexingService.processEvents(chainId);
>     });
> ```

I believe its because Ponder tries to reconcile ordering for events across chains
e.g. if chain 1 has an event at 0 seconds, chain 2 at 6 seconds, and chain 1 again at 12 seconds, Ponder should always execute them in the same, correct order, as you can see though this becomes a problem for fast chains which have sub 1 second block times.

**Comment by 0xOlias** [comment]
Fixed in #470 !
